src/quota_aggregator_impl.cc

```diff
@@ -210,25 +214,19 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
 
 // OnCacheEntryDelete will be called behind the cache_mutex_
 // no need to consider locking at this point
-//
-// if the element is refreshing and aggregated while waiting for the
-// response, tokens should be aggregated
 void QuotaAggregatorImpl::OnCacheEntryDelete(CacheElem* elem) {
-  if (elem->in_flight() && in_flush_all_ == false) {
-    // keep the cached entry remain in the cache if the entry is waiting for
-    // the response
-    cache_->Insert(elem->signature(), elem, 1);
-    return;
-  }
-
-  if (elem->is_aggregated() == true && in_flush_all_ == false) {
-    // Insert the element back to the cache while aggregator is waiting for the
-    // response.
-    elem->set_in_flight(true);
+  if (in_flush_all_ == false) {
+    // insert the element back to the cache
     cache_->Insert(elem->signature(), elem, 1);
 
-    AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
-        service_name_, service_config_id_));
+    if (elem->in_flight() == false) {
+      // refresh quota when the response is negative or aggregated.
+      if(elem->is_positive_response() == false || elem->is_aggregated()) {
+        elem->set_in_flight(true);
+        AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
```

Comment:
This is not good.   Basically,  we are requesting 1 token in every second even there is not traffic.
I think  "ping server for token" in CLOSE mode should be triggered by traffic, not by timer.  It should be traffic + timer.      

When a request come,  check last_ping_time,   if > 1second,  send another ping.